src/main/scala/org/broadinstitute/monster/etl/encode/EncodeExtractions.scala

```diff
@@ -0,0 +1,216 @@
+    }
+
+    override protected def newClient(): OkHttpClient = client
+  }
+
+  /**
+    * Pipeline stage which maps batches of search parameters into JSON entities
+    * from ENCODE matching those parameters.
+    *
+    * @param encodeEntity the type of ENCODE entity the stage should query
+    */
+  def getEntities(
+    encodeEntity: EncodeEntity
+  ): SCollection[List[(String, String)]] => SCollection[JsonObject] =
+    _.transform(s"Download ${encodeEntity.entryName} Entities") {
+      _.applyKvTransform(ParDo.of(new EncodeLookup(encodeEntity))).flatMap { kv =>
+        kv.getValue.fold(
+          throw _,
+          value => {
+            val decoded = for {
+              // TODO change to upack and msg instead of circe and json, dear future monster
+              json <- io.circe.parser.parse(value)
+              cursor = json.hcursor
+              objects <- cursor.downField("@graph").as[Vector[JsonObject]]
+            } yield {
+              objects
+            }
+            decoded.fold(throw _, identity)
+          }
+        )
+      }
+    }
+
+  /**
+    * Pipeline stage which extracts IDs from downloaded JSON entities for
+    * use in subsequent queries.
+    *
+    * @param entryName display name for the type of entity whose IDs will
+    *                  be extracted in this stage
+    * @param referenceField field in the input JSONs containing the IDs
+    *                       to extract
+    * @param manyReferences whether or not `referenceField` is an array
+    */
+  def getIds(
+    entryName: String,
+    referenceField: String,
+    manyReferences: Boolean
+  ): SCollection[JsonObject] => SCollection[String] =
+    _.transform(s"Get $entryName IDs") { collection =>
+      collection.flatMap { jsonObj =>
+        jsonObj(referenceField).toIterable.flatMap { referenceJson =>
+          val references = for {
+            refValues <- if (manyReferences) {
+              referenceJson.as[List[String]]
+            } else {
+              referenceJson.as[String].map { reference =>
+                List(reference)
+              }
+            }
+          } yield {
+            refValues
+          }
+          references.toOption
+        }.flatten
+      }.distinct
+    }
+
+  /**
+    * Pipeline stage which maps entity IDs into corresponding JSON entities
+    * downloaded from ENCODE.
+    *
+    * @param encodeEntity the type of ENCODE entity the input IDs correspond to
+   *  @param fieldName the field name to get the entity by, is "@id" by default
+    */
+  def getEntitiesByField(
+    encodeEntity: EncodeEntity,
+    fieldName: String = "@id"
+  ): SCollection[String] => SCollection[JsonObject] = { idStream =>
+    val paramsBatchStream =
+      idStream.transform(s"Build ${encodeEntity.entryName} ID Queries") {
+        _.map(KV.of("key", _))
+          .setCoder(KvCoder.of(StringUtf8Coder.of(), StringUtf8Coder.of()))
+          .applyKvTransform(GroupIntoBatches.ofSize(100))
```

Comment:
Do you think it's worth making the batch size an input parameter?

Reply:
Probably; input parameter from the command line, right?

Reply:
Yes, another command-line arg. You should be able to set a default of 100 to keep the default behavior the same